File: c/questions/deleteDuplicates.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* deleteDuplicates(struct ListNode* head) {
    struct ListNode dummy;
    dummy.next = head;
    struct ListNode* pre = &dummy;
    struct ListNode* cur = head;
    struct ListNode* dns;
    while (cur) {
        if (cur->next && cur->val==cur->next->val) {
            dns = pre;
            while (cur->next && cur->val==cur->next->val) {
                pre = cur;
                cur = cur->next;
                free(pre);
            }
            pre = cur;
            cur = cur->next;
            free(pre);
            dns->next = cur;
            pre = dns;
        } else {
            pre = cur;
            cur = cur->next;
        }
    }
    return dummy.next;
}

File: python/questions/deleteDuplicates.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        pre = dummy
        cur = head
        while cur:
            if cur.next and cur.val == cur.next.val:
                while cur.next and cur.val==cur.next.val:
                    cur = cur.next
                cur = cur.next
                pre.next = cur
            else:
                pre = cur
                cur = cur.next
        return dummy.next
